src/common/tse.c

```diff
@@ -925,14 +935,49 @@ tse_task_create(tse_task_func_t task_func, tse_sched_t *sched, void *priv,
 	return 0;
 }
 
+/* Insert dtp to the sleeping list of dsp. */
+static void
+tse_task_insert_sleeping(struct tse_task_private *dtp,
+			 struct tse_sched_private *dsp)
+{
+	struct tse_task_private *t;
+
+	if (d_list_empty(&dsp->dsp_sleeping_list)) {
+		d_list_add_tail(&dtp->dtp_list, &dsp->dsp_sleeping_list);
+		return;
+	}
+
+	/* If this task < the head, we don't need to search the list. */
+	t = d_list_entry(dsp->dsp_sleeping_list.next, struct tse_task_private,
+			 dtp_list);
+	if (dtp->dtp_wakeup_time < t->dtp_wakeup_time) {
+		/* Insert before the head. */
+		d_list_add(&dtp->dtp_list, &dsp->dsp_sleeping_list);
+		return;
+	}
+
+	/*
+	 * Search from the tail. Because this task >= the head, the search must
+	 * have a hit.
+	 */
+	d_list_for_each_entry_reverse(t, &dsp->dsp_sleeping_list, dtp_list) {
+		if (t->dtp_wakeup_time <= dtp->dtp_wakeup_time) {
+			/* Insert after t. */
+			d_list_add(&dtp->dtp_list, &t->dtp_list);
+			return;
+		}
+	}
+	D_ASSERT(false);
+}
+
 int
-tse_task_schedule(tse_task_t *task, bool instant)
+tse_task_schedule(tse_task_t *task, bool instant, uint64_t delay)
```

Comment:
this is a public API and will break existing libraries using it (same for reinit).
maybe create a new API tse_task_set_delay() that delays a task for x seconds after being scheduled?

Reply:
Sounds good to me. Will do.

Reply:
I added tse_task_schedule_with_delay and tse_task_reinit_with_delay instead, because a tse_task_set_delay could not be called at certain times (e.g., after the task has been put in the sleeping queue sorted by wakeup_time).